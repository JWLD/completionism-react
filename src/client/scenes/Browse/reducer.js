import { ADD_CATEGORY_DATA } from 'constants/action_types'

const initialState = {}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_CATEGORY_DATA:
      return { ...state, [action.category]: action.payload }
    default:
      return state
  }
}
