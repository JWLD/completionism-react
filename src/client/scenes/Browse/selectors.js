import { createSelector } from 'reselect'

export const categoryParamSelector = (state, props) =>
  props.match.params.category

const browseStateSelector = state => state.browse

export const categoryDataExistsSelector = createSelector(
  [browseStateSelector, categoryParamSelector],
  (browseState, categoryParam) => {
    return browseState[categoryParam] ? true : false
  }
)

export const categoryDataSelector = createSelector(
  [browseStateSelector, categoryParamSelector],
  (browseState, categoryParam) => browseState[categoryParam] || []
)
