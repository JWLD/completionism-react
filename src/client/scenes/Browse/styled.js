import styled from 'styled-components'

import { mixins, vars } from 'style'

export const BrowsePage = styled.div`
  ${mixins.page};
`

export const FilterWrap = styled.div`
  ${mixins.itemListWrap};

  position: relative;
`

export const FilterInput = styled.input`
  ${mixins.itemListItem};

  padding: 1rem;
  background-color: #eee;
  font: 2.5rem DinMC;
`

export const ClearInputBtn = styled.button`
  position: absolute;
  right: 0;
  height: ${vars.sizing.itemHeight}rem;
  width: ${vars.sizing.itemHeight}rem;
  margin-right: ${vars.sizing.itemListPadding}rem;
  color: gray;
  background-color: transparent;
  font-size: 3.25rem;
`
