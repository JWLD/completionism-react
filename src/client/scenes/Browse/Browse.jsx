import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import FaTimesCircle from 'react-icons/lib/fa/times-circle';

import NavBar from 'components/NavBar/NavBar';

import ItemList from './components/ItemList/ItemList';
import { fetchCategoryData } from './data/actions';
import './Browse.scss';

class Browse extends Component {
	constructor(props) {
		super(props);

		this.state = {
			category: this.props.match.params.category,
			filter: ''
		};

		this.onInputChange = this.onInputChange.bind(this);
		this.clearSearch = this.clearSearch.bind(this);
	}

	componentDidMount() {
		if (!this.props[this.state.category]) {
			this.props.fetchCategoryData(this.state.category);
		}
	}

	onInputChange(e) {
		this.setState({ filter: e.target.value });
	}

	clearSearch() {
		this.setState({ filter: '' });
	}

	render() {
		const iconClass = this.state.filter ? '' : 'hide';

		return (
			<div className="category-page">
				<NavBar />

				<div className="filter-box">
					<input onChange={this.onInputChange} value={this.state.filter} placeholder="Filter" />
					<FaTimesCircle onClick={this.clearSearch} className={iconClass} />
				</div>

				<ItemList
					category={this.state.category}
					content={Number(this.props.match.params.content)}
					categoryData={this.props[this.state.category] || []}
					filterVal={this.state.filter}
					routeProps={this.props.match.params}
				/>
			</div>
		);
	}
}

Browse.propTypes = {
	fetchCategoryData: PropTypes.func.isRequired,
	match: PropTypes.object.isRequired
};

const mapStateToProps = (state, ownProps) => ({
	[ownProps.match.params.category]: state.browse[ownProps.match.params.category]
});

const mapDispatchToProps = {
	fetchCategoryData,
};

export default connect(mapStateToProps, mapDispatchToProps)(Browse);
