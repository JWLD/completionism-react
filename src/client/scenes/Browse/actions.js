import * as ACTIONS from 'constants/action_types'
import { CONTENT_ARR } from 'constants/content'
import { fetchDBCategoryData } from 'services/api'
import history from 'lib/history'
import { activeCategorySelector, activeContentSelector } from 'Browse/selectors'

export const changeActiveCategory = category => ({
  type: ACTIONS.UPDATE_ACTIVE_CATEGORY,
  category
})

export const changeActiveContent = content => ({
  type: ACTIONS.UPDATE_ACTIVE_CONTENT,
  content
})

export const loadCategoryData = (categoryData, category) => ({
  type: ACTIONS.LOAD_CATEGORY_DATA,
  payload: categoryData,
  category
})

export const setActiveItemId = itemId => ({
  type: ACTIONS.SET_ACTIVE_ITEM_ID,
  payload: itemId
})

// THUNKS

export const changeBrowsePage = ({ next }) => (_, getState) => {
  const state = getState()
  const category = activeCategorySelector(state)
  const currentPage = activeContentSelector(state)

  const lastPage = CONTENT_ARR.length

  const newPage = next
    ? currentPage === lastPage ? 1 : currentPage + 1
    : currentPage === 1 ? lastPage : currentPage - 1

  history.push(`/browse/${category}/${newPage}`)
}

export const fetchCategoryData = category => dispatch => {
  return fetchDBCategoryData(category).then(categoryData => {
    return dispatch(loadCategoryData(categoryData, category))
  })
}
