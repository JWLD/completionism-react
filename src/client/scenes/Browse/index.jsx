import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { changeActiveCategory, setActiveContent } from 'Browse/actions'
import { categoryParamSelector, contentParamSelector } from 'Browse/selectors'
import { changeBrowsePage } from 'Browse/actions'

import * as SC from 'Browse/styled'
import NavBar from 'NavBar'
import FilterBox from 'FilterBox'
import ItemList from 'ItemList'
import SidePanel from 'SidePanel'

export class Browse extends Component {
  componentWillMount() {
    this.props.changeActiveCategory(this.props.category)
    this.props.setActiveContent(this.props.content)

    document.addEventListener('keydown', this.handleKeyDown)
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyDown)
  }

  handleKeyDown = ({ key }) => {
    if (key === 'ArrowLeft') {
      this.props.changeBrowsePage({ next: false })
    } else if (key === 'ArrowRight') {
      this.props.changeBrowsePage({ next: true })
    }
  }

  render() {
    return (
      <SC.BrowsePage>
        <NavBar />

        <SC.ListWrap>
          <FilterBox />
          <ItemList />
        </SC.ListWrap>
        <SidePanel />
      </SC.BrowsePage>
    )
  }
}

Browse.propTypes = {
  category: PropTypes.string.isRequired,
  changeActiveCategory: PropTypes.func.isRequired,
  changeBrowsePage: PropTypes.func.isRequired,
  content: PropTypes.number.isRequired,
  setActiveContent: PropTypes.func.isRequired
}

export const mapStateToProps = (state, ownProps) => ({
  category: categoryParamSelector(state, ownProps),
  content: contentParamSelector(state, ownProps)
})

const mapDispatchToProps = {
  changeActiveCategory,
  changeBrowsePage,
  setActiveContent
}

export default connect(mapStateToProps, mapDispatchToProps)(Browse)
