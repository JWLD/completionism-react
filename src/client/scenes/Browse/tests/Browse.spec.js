import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { Browse, mapStateToProps } from 'Browse'
import * as MOCKS from 'fixtures'

Enzyme.configure({ adapter: new Adapter() })

const mockFetchCategoryData = jest.fn()
let mockProps = {}

beforeEach(() => {
  mockProps = {
    category: '',
    categoryDataExists: false,
    fetchCategoryData: mockFetchCategoryData,
    ...MOCKS.BROWSE_ROUTER_PROPS
  }

  jest.resetAllMocks()
})

describe('<Browse />', () => {
  it('renders correctly', () => {
    const component = shallow(<Browse {...mockProps} />)

    expect(component).toMatchSnapshot()
  })
})

describe('#componentDidMount', () => {
  it('calls [fetchCategoryData] is [categoryDataExists] is false', () => {
    shallow(<Browse {...mockProps} />)

    expect(mockFetchCategoryData).toHaveBeenCalledTimes(1)
  })

  it('does not call [fetchCategoryData] when categoryData is populated', () => {
    mockProps.categoryDataExists = true
    shallow(<Browse {...mockProps} />)

    expect(mockFetchCategoryData).toHaveBeenCalledTimes(0)
  })
})

describe('#mapStateToProps', () => {
  it('extracts the required data from state', () => {
    const result = mapStateToProps(MOCKS.INITIAL_STATE, mockProps)

    expect(Object.keys(result)).toHaveLength(2)
    expect(result).toHaveProperty('category')
    expect(result).toHaveProperty('categoryDataExists')
  })
})
