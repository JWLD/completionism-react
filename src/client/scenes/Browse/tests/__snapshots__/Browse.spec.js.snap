// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Browse /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Browse
    category=""
    categoryDataExists={false}
    changeActiveCategory={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
      }
    }
    fetchCategoryData={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "category": "mounts",
          "content": 1,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <withRouter(NavBar) />,
        <ListWrap>
          <Connect(FilterBox) />
          <withRouter(Connect(ItemList)) />
        </ListWrap>,
        <withRouter(Connect(SidePanel)) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Connect(FilterBox) />,
            <withRouter(Connect(ItemList)) />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <withRouter(NavBar) />,
          <ListWrap>
            <Connect(FilterBox) />
            <withRouter(Connect(ItemList)) />
          </ListWrap>,
          <withRouter(Connect(SidePanel)) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Connect(FilterBox) />,
              <withRouter(Connect(ItemList)) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
